Fix copying of symlinks in containers

jlhawn opened the pull request:
Copying a symlink should do just that: copy the symlink NOT copy the target of the symlink. Also, the resulting file from the copy should have the name of the symlink NOT the name of the target file.

This patch includes a fix to _StatPath_, _ArchivePath_, and _ExtractToDir_ methods on containers. These operations _should_ resolve symlinks that are in the path but if the resource itself is a symlink then it _should not_ be resolved. This patch puts this logic into a common function `resolvePath` which resolves symlinks of the path's dir in scope of the container rootfs but does not resolve the final element of the path. Now archive, extract, and stat operations will return symlinks if the path is indeed a symlink.

Notably, destination paths should always be fully resolved, i.e., copying to a symlink should affect the symlink target, not the symlink itself.

jlhawn commented:
ping @tonistiigi @thaJeztah

jlhawn commented:
see discussion in https://github.com/docker/docker/pull/13171#issuecomment-123805210

icecrime commented:
Can you please let us know if this is strictly required for 1.8.0, and if we're only fixing usability issues or potential security holes?

jlhawn commented:
@icecrime So far it's just a usability/behavior issue. We haven't yet determined if there's a real security issue with this.

thaJeztah commented:
Because none of us is a security expert, I'd prefer to have the security team to have a look,
both at this change and the issue it should resolve (reported here https://github.com/docker/docker/pull/13171#issuecomment-123805210 and discussed on IRC, starting here: https://botbot.me/freenode/docker-dev/2015-07-22/?msg=45272608&page=6).
Better safe than sorry, now that we are still able to check before the 1.8 release.

ping @ewindisch @diogomonica @NathanMcCauley 

(just my 0.02, apologies if I crossed my boundaries here)

tonistiigi commented:
This seems to only address the Stat-header issue and not rebasing. If I run `docker cp id:/foo/ bar` where `foo` is a symlink I don't get local folder `bar` but a folder that has the name of the symlink target.

This breaks symlink targets without slash. `cp -a` seems to use same behavior for symlinks-to-dirs as regular directories. I think for local paths this was already different before but for remotes it changed with this PR.


Any comment on the other points in https://github.com/docker/docker/pull/13171#issuecomment-123805210 ?

tonistiigi commented on daemon/archive.go:
```diff
@@ -70,6 +70,31 @@ func (daemon *Daemon) ContainerExtractToDir(name, path string, noOverwriteDirNon
+func (container *Container) resolvePath(path string) (resolvedPath, absPath string, err error) {
+	// Consider the given path as an absolute path in the container.
+	absPath = path
+	if !filepath.IsAbs(absPath) {
```
I think we should do this clean step always. There's a breakout with path `/..` for example.

jlhawn replied:
ah, yeah. I'll remove this condition then.

icecrime commented:
Ok! Adding this to the milestone, we don't have to block 1.8.0-rc1 on it though.

jlhawn commented:
@tonistiigi 

> This seems to only address the Stat-header issue and not rebasing. If I run docker cp id:/foo/ bar where foo is a symlink I don't get local folder bar but a folder that has the name of the symlink target.

I'll add a test case for this and make sure it's fixed.

jlhawn commented on pkg/archive/archive.go:
```diff
@@ -37,11 +37,13 @@ type (
 		Compression      Compression
 		NoLchown         bool
 		ChownOpts        *TarChownOptions
-		Name             string
```
Rationale for removing this: It is never used, we have logic that uses it below, but I replaced it with the `rebaseNames` map.

Commit:
Fix `docker cp` Behavior With Symlinks

[pkg/archive] Update archive/copy path handling

  - Remove unused TarOptions.Name field.
  - Add new TarOptions.RebaseNames field.
  - Update some of the logic around path dir/base splitting.
  - Update some of the logic behind archive entry name rebasing.

[api/types] Add LinkTarget field to PathStat

[daemon] Fix stat, archive, extract of symlinks

  These operations *should* resolve symlinks that are in the path but if the
  resource itself is a symlink then it *should not* be resolved. This patch
  puts this logic into a common function `resolvePath` which resolves symlinks
  of the path's dir in scope of the container rootfs but does not resolve the
  final element of the path. Now archive, extract, and stat operations will
  return symlinks if the path is indeed a symlink.

[api/client] Update cp path hanling

[docs/reference/api] Update description of stat

  Add the linkTarget field to the header of the archive endpoint.
  Remove path field.

[integration-cli] Fix/Add cp symlink test cases

  Copying a symlink should do just that: copy the symlink NOT
  copy the target of the symlink. Also, the resulting file from
  the copy should have the name of the symlink NOT the name of
  the target file.

  Copying to a symlink should copy to the symlink target and not
  modify the symlink itself.

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

jlhawn commented:
A couple of `windows` test failures as expected... wish I could run those locally :-(.

thaJeztah commented:
@jlhawn I'm sure @ahmetalpbalkan is able to give you access to an azure machine :smile:

jlhawn commented:
@thaJeztah eh, no biggie... I have a windows 8.1 VM locally that I could use, I just haven't bothered setting up the docker dev environment on it. The test failure was also pretty minor: I was comparing a local windows backslash-delimited path with a constant unix-style forward-slash-delimited path.

jlhawn commented:
@tonistiigi This should now fix the behavior you saw on the source being a symlink to a directory.

As for your other question from https://github.com/docker/docker/pull/13171#issuecomment-123805210:

> How can we protect this from running containers updating a filesystem/volumes when the request is taking place? I mean we check for breakouts in the beginning of GET/PUT but if a container is using the filesystem it could just flip a symlink in the right time and then we would have full read/write to host.

I think the only thing we could do is require that the container be stopped or paused? (can that be considered outside the scope of this PR?)

tonistiigi commented:
Is `docker cp` with symlinks supposed to follow the semantics of `cp -a`? Docs are bit contradictory saying it's `cp -a` but also that symlink is always/only followed if it's followed by a slash.

```
#setup (GNU coreutils 8.21):
mkdir foo
touch foo/bar
ln -s foo baz
mkdir abc
touch abc/def

cp -a abc baz # creates foo/abc/def
docker cp t0:abc baz # error: cannot copy directory
docker cp abc t0:baz # error: cannot copy directory

cp -a abc/def baz # creates foo/def
docker cp t0:abc/def baz # creates baz
docker cp abc/def t0:baz # creates baz

cp -a baz/. abc # creates abc/bar
docker cp t0:baz/. abc # creates abc/bar
docker cp baz/. t0:abc # creates abc/(baz -> foo) + WARN
```


The `resolvePath()` logic itself seems safe but I'm bit concerned about the readability of the actual place where the breakout protection happens. There is bunch of logic in `TarResource` for testing `symlink/` paths but if I understand it correctly this is only used in client side and would be unsafe otherwise. Maybe we could add a comment in the daemon code that calls `TarResource` that the path sent there can be `symlink` or `dir/` but should never be `symlink/` or `symlink/.` and it's the calling code's responsibility to check that.

Having two separate code paths that do tar rebasing probably also isn't ideal maintenance-wise and I can't help to point out that by not having the first-level directory we could have probably avoided the need to do rebasing in both ends(plus the need to validate downloaded tar that we should probably also do).

tonistiigi commented on daemon/archive.go:
```diff
@@ -257,12 +241,6 @@ func (container *Container) ExtractToDir(path string, noOverwriteDirNonDir bool,
-		return err
-	}
-	absPath = filepath.Join("/", baseRel)
-
```
This does not seem right. If the path is `/myvolume/symlinktoroot/` it would be matched to the volume and writing would go to the readonly rootfs.

jlhawn commented:
> Is `docker cp` with symlinks supposed to follow the semantics of `cp -a`? Docs are bit contradictory saying it's `cp -a` but also that symlink is always/only followed if it's followed by a slash.

This should only apply to the source argument. From your example (thanks!) it looks like we've also erroneously applied it to the destination argument :frowning:. I'll fold this into this PR also (and add a test case for it based on your example).

jlhawn commented:
Okay, all updated again. Currently the Windows CI is on fire, so I guess we'll have to wait on that.

tonistiigi commented on daemon/archive.go:
```diff
@@ -154,41 +180,33 @@ func (container *Container) ArchivePath(path string) (content io.ReadCloser, sta
 
 	stat = &types.ContainerPathStat{
-		Name:  lstat.Name(),
+		Name:  filepath.Base(absPath),
```
What's the definition of `Name`? If its always basename of `Path` then why return as a separate field at all?

jlhawn replied:
Yes, it's the basename, while `Path` is the absolute path.

jlhawn replied:
I'll go ahead ad remove the `Path` field. Your right, it's unnecessary given that the user has to provide a path in the first place.

tonistiigi replied:
Well, `Path` is cleaned and made absolute so it may actually have some value. `Name`, on the other hand, anyone can easily get from `Path` (if the definition remains as it is currently).

tonistiigi commented on daemon/archive.go:
```diff
@@ -154,41 +180,33 @@ func (container *Container) ArchivePath(path string) (content io.ReadCloser, sta
 		Path:  absPath,
 		Size:  lstat.Size(),
 		Mode:  lstat.Mode(),
 		Mtime: lstat.ModTime(),
```
No `LinkTarget` here? GET and HEAD should match.

jlhawn replied:
Thanks for the catch! I'll move this into a separate method so that logic will only exist in one place.

tonistiigi commented on daemon/archive.go:
```diff
@@ -227,27 +245,21 @@ func (container *Container) ExtractToDir(path string, noOverwriteDirNonDir bool,
 	}
 
+	// The destination path needs to be resolved to a host path, with all
+	// sybolic links followed in the scope of the container's rootfs. Note
```
symbolic

tonistiigi commented on daemon/archive.go:
```diff
@@ -257,12 +269,6 @@ func (container *Container) ExtractToDir(path string, noOverwriteDirNonDir bool,
-		return err
-	}
-	absPath = filepath.Join(string(os.PathSeparator), baseRel)
-
```
The `/volume/symlinktoroot` pattern does not seem to be handled yet.

```
docker run --read-only=true -v /vol --name t1 busybox ln -s .. /vol/root
docker cp hello t1:/vol/root/ # should fail
```

jlhawn replied:
Thanks! I'll undelete this and add add a test case specifically for this.

jlhawn replied:
ooh, this is actually a bit tricky to test using `docker cp` so I'll have to test it using the API directly.

tonistiigi commented:
LGTM

icecrime commented:
LGTM

calavera commented on integration-cli/docker_cli_cp_test.go:
```diff
@@ -250,29 +250,189 @@ func (s *DockerSuite) TestCpAbsoluteSymlink(c *check.C) {
+	if err := os.MkdirAll(filepath.Join(testDir, cpTestPath), os.FileMode(0700)); err != nil {
+		c.Fatal(err)
+	}
+	hostFile, err := os.Create(filepath.Join(testDir, cpFullPath))
```
`ioutil.WriteFile`?

jlhawn replied:
updated!

calavera commented:
LGTM, but can we squash all those commits? all of the are related to the same thing and it would be easier to keep track of the for the release.

jlhawn commented:
@calavera okay, I'll squash them together

calavera commented:
:green_heart: Merging!

jlhawn commented:
:tada:

calavera commented:
cherry picked into #15091

thaJeztah commented:
Great work guys!
